opentracing/dapperish/main.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"html"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/opentracing/api-golang/opentracing"
+)
+
+func client() {
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		span, ctx := opentracing.StartSpan("getInput")
+		span.Info("ctx: ", ctx)
+		fmt.Print("Enter text: ")
+		text, _ := reader.ReadString('\n')
+		span.Info(text)
+
+		httpClient := &http.Client{}
+		httpReq, _ := http.NewRequest("POST", "http://localhost:8080/", bytes.NewReader([]byte(text)))
+		opentracing.AddContextIDToHttpHeader(span.ContextID(), httpReq.Header)
+		resp, err := httpClient.Do(httpReq)
+		if err != nil {
+			span.Error("error: ", err)
+		} else {
+			span.Info("got response: ", resp)
+		}
+
+		span.Finish()
+	}
+}
+
+func server() {
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		reqCtxID, err := opentracing.GetContextIDFromHttpHeader(
+			req.Header, opentracing.GlobalTracer())
+		if err != nil {
+			panic(err)
+		}
+
+		serverSpan, _ := opentracing.StartSpan("serverSpan", reqCtxID)
```

Comment:
First of all, I agree completely that causality should be described at higher levels of the API, not just in raw annotations.

Anyway, I'm imagining that the client (i.e., the hyperbahn client) could do this... if we model each hop as two spans – one on the client, one in some routing server – obviously each server span is a blocking child of the corresponding client span. The trick is to describe the causal relationships between the various (hop1, hop2, etc) client-side spans. One possibility would be to add a `BlockedOnCompletion` annotation that points to a previously-completed `ContextId`; if that's too much bookkeeping, then the [hyperbahn] client can merely emit special `Log` statements that explicitly specify that future span children depend on previously completed span children (without actually specifying span_ids or similar for those children). I am thinking this would look vaguely like `Span.AddCausalityHint(opentracing.FutureChildrenDependOnPreviouslyCompletedChildren)`, though hopefully with a more concise name!

Not sure if the above is clear or not.